config/config_test.go

```diff
@@ -0,0 +1,205 @@
+			mode:             os.FileMode(0600),
+			expectedWarnings: 0,
+		},
+		{
+			name:             "group-readable and world-readable",
+			mode:             os.FileMode(0644),
+			expectedWarnings: 2,
+		},
+		{
+			name:             "group-readable",
+			mode:             os.FileMode(0640),
+			expectedWarnings: 1,
+		},
+		{
+			name:             "world-readable",
+			mode:             os.FileMode(0604),
+			expectedWarnings: 1,
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			assert := assert.New(t)
+
+			dir, err := os.MkdirTemp("", "rancher-cli-test-*")
+			assert.NoError(err)
+			defer os.RemoveAll(dir)
+
+			path := filepath.Join(dir, "cli2.json")
+			err = os.WriteFile(path, []byte(validFile), tt.mode)
+			assert.NoError(err)
+
+			warnings, err := GetFilePermissionWarnings(path)
+			assert.NoError(err)
+			assert.Len(warnings, tt.expectedWarnings)
+		})
+	}
+}
+
+func Test_Permission(t *testing.T) {
+	t.Parallel()
+
+	// New config files should have 0600 permissions
+	t.Run("new config file", func(t *testing.T) {
+		t.Parallel()
+		assert := assert.New(t)
+
+		dir, err := os.MkdirTemp("", "rancher-cli-test-*")
+		assert.NoError(err)
+		defer os.RemoveAll(dir)
+
+		path := filepath.Join(dir, "cli2.json")
+		conf, err := LoadFromPath(path)
+		assert.NoError(err)
+
+		err = conf.Write()
+		assert.NoError(err)
+
+		info, err := os.Stat(path)
+		assert.NoError(err)
+		assert.Equal(os.FileMode(0600), info.Mode())
+
+		// make sure new file doesn't create permission warnings
+		warnings, err := GetFilePermissionWarnings(path)
+		assert.NoError(err)
+		assert.Len(warnings, 0)
+	})
+	// Already existing config files should keep their current permissions
+	t.Run("existing config file", func(t *testing.T) {
+		t.Parallel()
+		assert := assert.New(t)
+
+		dir, err := os.MkdirTemp("", "rancher-cli-test-*")
+		assert.NoError(err)
+		defer os.RemoveAll(dir)
+
+		path := filepath.Join(dir, "cli2.json")
+		err = os.WriteFile(path, []byte(validFile), 0700)
+		assert.NoError(err)
+
+		conf, err := LoadFromPath(path)
+		assert.NoError(err)
+
+		err = conf.Write()
+		assert.NoError(err)
+
+		info, err := os.Stat(path)
+		assert.NoError(err)
+		assert.Equal(os.FileMode(0700), info.Mode())
+	})
+}
+
+func Test_LoadFromPath(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name         string
+		content      string
+		expectedConf Config
+		expectedErr  bool
+	}{
+		{
+			name:    "valid config",
+			content: validFile,
+			expectedConf: Config{
+				Servers: map[string]*ServerConfig{
+					"rancherDefault": {
+						AccessKey: "the-access-key",
+						SecretKey: "the-secret-key",
+						TokenKey:  "the-token-key",
+						URL:       "https://example.com",
+						Project:   "cluster-id:project-id",
+						CACerts:   "",
+					},
+				},
+				CurrentServer: "rancherDefault",
+			},
+		},
+		{
+			name:        "invalid config",
+			content:     invalidFile,
+			expectedErr: true,
```

Comment:
Nit: This function has a bit of an odd output. In the cases of an error, it still returns the default config file. Since this function is exported, I think that this should be something that we assert on, since the caller could choose to log the error and do some other processing with the default value returned alongside the error.

Reply:
I might be misunderstanding so apologies in advance.

I believe if `LoadFromPath` returns an error, the config return shouldn't be used, so just good old idiomatic go error handling. However, I see that currently we're not returning an empty config, we're returning a default valid one with the path set.

I think it would make sense to make `LoadFromPath` return an empty config on error, that empty config shouldn't be used, so we don't really need to assert it.

So:
- File already exists: return the config, nil
- File doesn't exist: return default config, nil
- Error reading file: return zero value config, err

Let me know.

Reply:
I have made the changed mentioned above, so I think we don't need to add an assert.